### PR TITLE
Use azd binding to create connections

### PR DESCRIPTION
**Pull request by brycechen1849**
This is a demo sample based on the nodejs-mongo-aca project, using new azd binding feature to create connections from host to service. 
The changed items are::
- Specify bindings in azure.yaml, this will store connection strings in app config, which further reference to keyvault.
- Project specific main.bicep is modified:
  - App config and keyvault will be uses MI + endpoint authentication specified in bicep
  - CosmosDB and App Insights will be connected via binding feature that writes the connection string into above app config and further reference to the keyvault
- Api server init config will read the configurations from app config (instead of host env or keyvault directly).

### CODE REVIEW

File: templates/todo/api/js/src/config/index.ts [hunk 1]
```diff
@@ -2,17 +2,30 @@ import { AppConfig, DatabaseConfig, ObservabilityConfig } from "./appConfig";
...
 import { SecretClient } from "@azure/keyvault-secrets";
+import { load } from "@azure/app-configuration-provider";
 import { logger } from "../config/observability";
 import { IConfig } from "config";
 
 export const getConfig: () => Promise<AppConfig> = async () => {
-    // Load any ENV vars from local .env file
-    if (process.env.NODE_ENV !== "production") {
+    logger.info(`Loading configuration in ${process.env.NODE_ENV} environment...`);
+    if (process.env.NODE_ENV == "production") {
```
Comment (yungezz) [inline templates/todo/api/js/src/config/index.ts:11]: ignore case string comparasion

File: templates/todo/api/js/src/config/index.ts [hunk 2]
```diff
@@ -68,4 +81,37 @@ const populateEnvironmentFromKeyVault = async () => {
...
+
+        const settings = await load(endpoint, credential, {
+            keyVaultOptions: {
+                // Access keyvault using the same identity, make sure the permission is set correctly.
+                credential: credential
+            }
+        });
+
+        for (const [key, value] of settings) {
+            logger.info(`Setting read from app config ${key}=${value}`);
```
Comment (yungezz) [inline templates/todo/api/js/src/config/index.ts:109]: pls not log config key value, especially value

File: templates/todo/api/js/src/config/index.ts [hunk 3]
```diff
@@ -68,4 +81,37 @@ const populateEnvironmentFromKeyVault = async () => {
...
+            process.env[key] = value;
+        }
+    }
+    catch (err: any) {
+        logger.error(`Error authenticating with Azure App Configuration or Keyvault. Ensure your managed identity or service principal has GET/LIST permissions. Error: ${err}`);
```
Comment (yungezz) [inline templates/todo/api/js/src/config/index.ts:114]: may not be auth error

File: templates/todo/api/js/package.json [hunk 4]
```diff
@@ -24,6 +24,7 @@
     ]
   },
   "dependencies": {
+    "@azure/app-configuration-provider": "1.0.0-preview.4",
```
Comment (houk-ms) [inline templates/todo/api/js/package.json:27]: any stable version?

File: templates/todo/projects/nodejs-mongo-aca/.repo/bicep/azure.yaml [hunk 5]
```diff
@@ -32,11 +43,11 @@ hooks:
...
-    posix:
-      shell: sh
-      run: 'rm ./src/web/.env.local'
+  # postdeploy:
+  #   windows:
+  #     shell: pwsh
+  #     run: 'rm ./src/web/.env.local'
+  #   posix:
+  #     shell: sh
+  #     run: 'rm ./src/web/.env.local'
```
Comment (houk-ms) [inline templates/todo/projects/nodejs-mongo-aca/.repo/bicep/azure.yaml:52]: why these are commented

File: templates/todo/api/js/src/config/index.ts (hunk 3 above)
Comment (brycechen1849) [reply templates/todo/api/js/src/config/index.ts]: fixed updating error message

File: templates/todo/api/js/package.json (hunk 4 above)
Comment (brycechen1849) [reply templates/todo/api/js/package.json:27]: No. latest still in preview https://github.com/Azure/AppConfiguration-JavaScriptProvider/tags

File: templates/todo/api/js/src/config/index.ts (hunk 1 above)
Comment (brycechen1849) [reply templates/todo/api/js/src/config/index.ts]: fixed with toLower()

File: templates/todo/api/js/src/config/index.ts (hunk 2 above)
Comment (brycechen1849) [reply templates/todo/api/js/src/config/index.ts]: Fixed removing log

File: templates/todo/projects/nodejs-mongo-aca/.repo/bicep/azure.yaml [hunk 6]
```diff
@@ -28,7 +39,7 @@ hooks:
   predeploy:
     windows:
       shell: pwsh
-      run: 'echo "VITE_API_BASE_URL=\"$env:API_BASE_URL\"" > ./src/web/.env.local ; echo "VITE_APPLICATIONINSIGHTS_CONNECTION_STRING=\"$env:APPLICATIONINSIGHTS_CONNECTION_STRING\"" >> ./src/web/.env.local'
+      run: 'echo "VITE_API_BASE_URL=$env:API_BASE_URL" > ./src/web/.env.local ; echo "VITE_APPLICATIONINSIGHTS_CONNECTION_STRING=$env:APPLICATIONINSIGHTS_CONNECTION_STRING" >> ./src/web/.env.local'
```
Comment (brycechen1849) [inline templates/todo/projects/nodejs-mongo-aca/.repo/bicep/azure.yaml:42]: I'm using windows 11 with PowerShell 7.4.2 (`pwsh --version` shows PowerShell 7.4.2) and it causes the .env.local to look like this when running the `echo` command:
```
VITE_API_BASE_URL=/
https://ca-api-zdntfchj4zeri.thankfulground-ec33f819.eastus.azurecontainerapps.io/
VITE_APPLICATIONINSIGHTS_CONNECTION_STRING=/
InstrumentationKey=cb775b3f-ca95-4b0d-9b4c-02d386bd39f9;IngestionEndpoint=https://eastus-8.in.applicationinsights.azure.com/;LiveEndpoint=https://eastus.livediagnostics.monitor.azure.com/;ApplicationId=38cb1ed5-58f0-4952-8f19-ce6b20597d9b/
```
So the bundled web container can't read connection appinsight string correctly and the web application in browser stuck at loading

Comment (jongio) [reply templates/todo/projects/nodejs-mongo-aca/.repo/bicep/azure.yaml:42]: @vhvb1989 - Thoughts on this need?

Comment (vhvb1989) [reply templates/todo/projects/nodejs-mongo-aca/.repo/bicep/azure.yaml:42]: this was fixed here: https://github.com/Azure/azure-dev/pull/3671/files

You might want to pull upstream

Comment (brycechen1849) [reply templates/todo/projects/nodejs-mongo-aca/.repo/bicep/azure.yaml]: fixed by merging the hook fix.

**Pull request closed without merging**